### PR TITLE
flamenco, fuzz: ELF loader fuzz target

### PR DESCRIPTION
**Pull request by ravyu-jump**

### CODE REVIEW

File: src/flamenco/runtime/tests/fd_exec_instr_test.h [hunk 1]
```diff
@@ -71,6 +71,13 @@ fd_exec_instr_test_run( fd_exec_instr_test_runner_t *        runner,
...
                         ulong                                output_bufsz );
 
+ulong
+fd_sbpf_program_load_test_run( void const *         _bin,
+                               ulong                elf_sz,
+                               fd_exec_test_elf_loader_effects_t ** output,
+                               void *                               output_buf,
+                               ulong                                output_bufsz );
+
 FD_PROTOTYPES_END
```
Comment (topointon-jump) [inline src/flamenco/runtime/tests/fd_exec_instr_test.h:81]: When this function is finished, could you add a comment like the one above?

File: src/flamenco/runtime/tests/fd_exec_test.proto [hunk 2]
```diff
@@ -111,3 +111,20 @@ message InstrFixture {
     InstrContext input = 1;
     InstrEffects output = 2;
 }
+
+message ELFLoaderEffects {
+    // loaded program rodata
```
Comment (topointon-jump) [inline src/flamenco/runtime/tests/fd_exec_test.proto:116]: Before merging, could you add a comment above explaining what this is for?

File: src/ballet/sbpf/fd_sbpf_loader.h [hunk 3]
```diff
@@ -183,7 +183,11 @@ fd_sbpf_program_new( void *                     prog_mem,
      reject_broken_elfs: true
 
    For documentation on these config params, see:
-   https://github.com/solana-labs/rbpf/blob/v0.3.0/src/vm.rs#L198 */
+   https://github.com/solana-labs/rbpf/blob/v0.3.0/src/vm.rs#L198 
+
+   Solana/Agave equivalent:
+   https://github.com/solana-labs/rbpf/blob/main/src/elf.rs#L361
+   */
```
Comment (topointon-jump) [inline src/ballet/sbpf/fd_sbpf_loader.h:190]: Minor nit: could your links to Agave source be pinned to a specific tag? Like the one above.

File: src/flamenco/runtime/tests/fd_exec_instr_test.c [hunk 4]
```diff
@@ -860,3 +863,105 @@ fd_exec_instr_test_run( fd_exec_instr_test_runner_t *        runner,
...
+  fd_exec_test_elf_loader_effects_t * elf_effects =
+    FD_SCRATCH_ALLOC_APPEND( l, alignof(fd_exec_test_elf_loader_effects_t),
+                                sizeof (fd_exec_test_elf_loader_effects_t) );
+  if( FD_UNLIKELY( _l > output_end ) ) {
+    /* return 0 on fuzz-specific failures */
+    return 0UL;
+  }
+  fd_memset( elf_effects, 0, sizeof(fd_exec_test_elf_loader_effects_t) );
+
+  do{
```
Comment (topointon-jump) [inline src/flamenco/runtime/tests/fd_exec_instr_test.c:909]: Mind commenting what this control-flow pattern is for? Thanks :)

**Review by topointon-jump — DISMISSED** [review]

File: src/ballet/sbpf/fd_sbpf_loader.c [hunk 5]
```diff
@@ -546,6 +545,12 @@ fd_sbpf_program_new( void *                     prog_mem,
     return NULL;
   }
 
+  /* https://github.com/solana-labs/rbpf/blob/v0.8.0/src/elf_parser/mod.rs#L99 */
+  if( FD_UNLIKELY( !fd_ulong_is_aligned( (ulong) rodata, 8UL ) ) ){
+    FD_LOG_WARNING(( "rodata is not 8-byte aligned" ));
+    return NULL;
+  }
```
Comment (topointon-jump) [inline src/ballet/sbpf/fd_sbpf_loader.c:557]: In some places in the BPF loaders we allocate rodata with 32-byte alignment, should we change these to be 8-byte aligned?

**Review by topointon-jump — DISMISSED** [review]
Nice!

File: src/flamenco/runtime/tests/fd_exec_instr_test.c [hunk 6]
```diff
@@ -871,15 +871,32 @@ fd_sbpf_program_load_test_run( fd_exec_test_elf_loader_ctx_t const * input,
...
   }
-  void const * _bin = input->elf.data->bytes;
+
+  /* Nanopb's pb_bytes_array adds a 4-byte "size" field to the ELF binary,
+    which causes the actual binary data to be misaligned when it is passed
+    to the ELF loader.
+
+    As a temporary workaround, the entire binary is copied to an 8-byte aligned 
+    region of memory. This ensures that the binary data starts at an 8-byte aligned
+    address, satisfying the alignment requirements of the ELF loader.
```
[the remark this answers is not in the document]
Comment (ravyu-jump) [reply src/flamenco/runtime/tests/fd_exec_instr_test.c:886]: yeah missed this one. Thanks!

File: src/ballet/sbpf/fd_sbpf_loader.c (hunk 5 above)
Comment (ravyu-jump) [reply src/ballet/sbpf/fd_sbpf_loader.c:557]: nice catch, free (best case) 24 bytes!!

**Review by topointon-jump — DISMISSED** [review]

**Review by topointon-jump — APPROVED** [review]

**Pull request merged**